renku/service/views/error_handlers.py

```diff
@@ -0,0 +1,403 @@
+
+        # NOTE: HTTPException are now handled in the entrypoint
+        except ServiceError as e:
+            return error_response(e)
+        # NOTE: GitError here may not be necessary anymore
+        except GitError as e:
+            return error_response(ProgramGitError(e, e.message if e.message else None))
+        except (Exception, BaseException, OSError, IOError) as e:
+            if hasattr(e, "stderr"):
+                error_message = " ".join(e.stderr.strip().split("\n"))
+            else:
+                error_message = str(e)
+
+            return error_response(ProgramInternalError(e, error_message))
+
+    return decorated_function
+
+
+def handle_common_except(f):
+    """Handle common exceptions."""
+    # noqa
+    @wraps(f)
+    def dec(*args, **kwargs):
+        """Decorated function."""
+
+        @handle_base_except
+        @handle_validation_except
+        @handle_renku_except
+        @handle_git_except
+        @handle_jwt_except
+        def _wrapped(*args_, **kwargs_):
+            return f(*args_, **kwargs_)
+
+        return _wrapped(*args, **kwargs)
+
+    return dec
+
+
+def handle_templates_read_errors(f):
+    """Wrapper which handles reading templates errors."""
+    # noqa
+    @wraps(f)
+    def decorated_function(*args, **kwargs):
+        """Represents decorated function."""
+        try:
+            return f(*args, **kwargs)
+        except InvalidTemplateError as e:
+            raise UserTemplateInvalidError(e)
+        except GitError as e:
+            error_message = str(e)
+            if "Cannot clone repo from" in error_message:
+                raise UserRepoUrlInvalidError(e)
+            elif "Cannot checkout reference" in error_message:
+                raise UserRepoBranchInvalidError(e)
+            elif "Cannot checkout manifest file" in error_message:
+                raise UserTemplateInvalidError(e)
+            raise
+
+    return decorated_function
+
+
+@handle_templates_read_errors
+def handle_templates_create_errors(f):
+    """Wrapper which handles template creating projects errors."""
+    # noqa
+    @wraps(f)
+    def decorated_function(*args, **kwargs):
+        """Represents decorated function."""
+
+        def match_schema(target, message):
+            if re.match(f".*{target}.*contains.*unsupported characters.*", message):
+                return True
+            return False
+
+        try:
+            return f(*args, **kwargs)
+        except ValidationError as e:
+            if getattr(e, "field_name", None) == "_schema":
+                error_message = str(e)
+                if match_schema("Project name", error_message):
+                    raise UserProjectCreationError(e, "project name must contain at least a valid character")
+                elif match_schema("git_url", error_message):
+                    raise ProgramProjectCreationError(e, "git_url is invalid")
+            raise
+        except KeyError as e:
```

Comment:
This is pretty bad, to be honest (not on your part, in general), whatever we decorate here should handle the KeyError and reraise it as a more descriptive error. It shouldn't be up to the caller to figure out what could have happened. We should create a follow-up issue to clean this up

Reply:
Yeah, this is a mix of different errors. In general, a good strategy could be trying to catch all the errors we can identify (especially if they are likely to be determined by wrong data provided by users), and let the others trigger a more generic `ProgramSomthingError` since it's always possible the client is passing wrong keys in creative ways.
If that sounds like a reasonable strategy, I'll open a follow-up issue with a few details.

Reply:
#2729